src/mbgl/source/source.cpp

```diff
@@ -473,6 +474,87 @@ void Source::updateTilePtrs() {
     }
 }
 
+vec2<int16_t> coordinateToTilePoint(const TileID& tileID, const TileCoordinate& coord) {
+    auto zoomedCoord = coord.zoomTo(tileID.sourceZ);
+    return {
+        int16_t(util::clamp<int64_t>((zoomedCoord.x - (tileID.x + tileID.w * std::pow(2, tileID.sourceZ))) * util::EXTENT,
+                    std::numeric_limits<int16_t>::min(),
+                    std::numeric_limits<int16_t>::max())),
+        int16_t(util::clamp<int64_t>((zoomedCoord.y - tileID.y) * util::EXTENT,
+                    std::numeric_limits<int16_t>::min(),
+                    std::numeric_limits<int16_t>::max()))
+    };
+}
+
+struct TileQuery {
+    Tile* tile;
+    GeometryCollection queryGeometry;
+    double tileSize;
+    double scale;
+};
+
+std::unordered_map<std::string, std::vector<std::string>> Source::queryRenderedFeatures(
+        const std::vector<TileCoordinate>& queryGeometry,
+        const double zoom,
+        const double bearing,
+        const optional<std::vector<std::string>>& layerIDs) {
+
+    std::unordered_map<std::string, std::vector<std::string>> result;
+
+    double minX = std::numeric_limits<double>::infinity();
+    double minY = std::numeric_limits<double>::infinity();
+    double maxX = -std::numeric_limits<double>::infinity();
+    double maxY = -std::numeric_limits<double>::infinity();
+    double z = queryGeometry[0].z;
+
+    for (auto& c : queryGeometry) {
+        minX = util::min(minX, c.x);
+        minY = util::min(minY, c.y);
+        maxX = util::max(maxX, c.x);
+        maxY = util::max(maxY, c.y);
+    }
+
+    std::unordered_map<uint64_t, TileQuery> tileQueries;
+
+    for (auto& tilePtr : tilePtrs) {
```

Comment:
The two loops and `TileQuery` objects are needed for queries that cross the international date line. These queries that cross the date line can intersect a tile in multiple spots. TileQuery` objects are used to collect these separate query geometry parts into a single multi-geometry query. A single multi-geometry query let's us avoid returning duplicates without having a separate dedupe step later.